### PR TITLE
ENH: Standardized timeseries accessor names, #9606

### PR DESCRIPTION
**Pull request by emilydolson**
In accordance with https://github.com/pydata/pandas/issues/9606, I added `day_of_week` and `day_of_year` accessors to DatetimeIndex, Timestamp, and Period. Since https://github.com/pydata/pandas/issues/9606 didn't mention `weekofyear`, and there's already `week` as a replacement for that, I didn't add a `week_of_year`, but I'm happy to.

I updated the tests and docs, but this is my first PR to pandas involving actual code, so let me know if I missed anything.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
For `dayofweek`/`day_of_week` we also already have `weekday`, so this is adding a third alias for that one .. 
So I am a bit torn on this one (although I like the fact that we try to standardize the namings)

**Comment by jreback** [comment]
@jorisvandenbossche I think we should deprecate the original ones in 0.18 anyhow (eg daysinmonth)

File: pandas/tseries/index.py [hunk 1]
```diff
@@ -1458,7 +1460,9 @@ def _set_freq(self, value):
     dayofweek = _field_accessor('dayofweek', 'dow',
                                  "The day of the week with Monday=0, Sunday=6")
     weekday = dayofweek
+    day_of_week = dayofweek
     dayofyear = _field_accessor('dayofyear', 'doy', "The ordinal day of the year")
+    day_of_year = dayofyear
     quarter = _field_accessor('quarter', 'q', "The quarter of the date")
```
Comment (jreback) [inline pandas/tseries/index.py]: let's switch the impl to the new names (and make the original ones aliases)
that way they are easy to remove

File: doc/source/api.rst [hunk 2]
```diff
@@ -499,8 +499,10 @@ These can be accessed like ``Series.dt.<property>``.
    Series.dt.week
    Series.dt.weekofyear
```
Comment (jreback) [inline doc/source/api.rst]: week_of_year as well

**Comment by jorisvandenbossche** [comment]
Maybe we could also remove these from the dir to avoid all the duplicates in tab completion?

**Comment by jreback** [comment]
there is a test in test_series for dt accessor that needs updating as well

**Comment by jreback** [comment]
yep @jorisvandenbossche agreed

File: doc/source/api.rst (hunk 2 above)
Comment (jorisvandenbossche) [inline doc/source/api.rst]: I think we should choose between `week` or `week_of_year` here. If we choose for `week` as the 'main' attribute, then I don't think it is needed to add `week_of_year`

But I don't really have a strong preference between `week` or `week_of_year`

**Comment by jorisvandenbossche** [comment]
I think choosing one prefered name per attribute, and remove all others from the docs (only mentioning them somewhere as being aliases) and tab completion, is maybe a less 'aggresive' way for striving for consistency instead of really deprecating them

**Comment by jreback** [comment]
sure that's fine too

**Comment by jreback** [comment]
@emilydolson you will need to define

def _dir_deletions()

which returns a set of the attributes to remove

**Comment by emilydolson** [comment]
I've got the tab completion fixed for DatetimeIndex and PeriodIndex, but defining `_dir_deletions()` in Timestamp doesn't seem to affect the return value of `__dir__()`. Is there something about Cython that might interfere with _dir_deletions?

**Comment by jreback** [comment]
so the issue here is that ``Timestamp`` needs to inherit from ``PandasObject``, which is in ``pandas.core.base``. But you can't simply include that as it would have a circular dependency.

So here is what we need to do.

Move from ``pandas.core.base``, ``StringMixin`` and ``PandasObject`` to a new module, call it ``pandas.core.object``. import this to ``pandas.core.base`` (you *may* want to change imports of ``PandasObject`` elsewhere in the codebase). Then you can import this as a mix-in for ``Timestamp`` (in tslib.pyx).

long-story-short this will then work.

**Comment by jreback** [comment]
can you rebase / update

also let's deprecate the old names (e.g. w/o the underscores).

You will need to put the warning in the accessor (and probably create a specific function/property for them in order to do this as they are defined by the ``_field_accessor`` ATM)

**Comment by jreback** [comment]
can you update

**Comment by jreback** [comment]
@emilydolson can you update according to comments

**Comment by jreback** [comment]
@emilydolson can you rebase/update according to comments

**Comment by jreback** [comment]
closing, but pls reopen if you'd like to update, which would be great.

**Pull request closed without merging**